include/grpcpp/impl/codegen/server_callback_impl.h

```diff
@@ -222,34 +215,55 @@ class ServerCallbackReaderWriter {
 };
 
 // The following classes are the reactor interfaces that are to be implemented
-// by the user, returned as the result of the method handler for a callback
-// method, and activated by the call to OnStarted. The library guarantees that
-// OnStarted will be called for any reactor that has been created using a
-// method handler registered on a service. No operation initiation method may be
-// called until after the call to OnStarted.
-// Note that none of the classes are pure; all reactions have a default empty
-// reaction so that the user class only needs to override those classes that it
-// cares about.
+// by the user, returned as the output parameter of the method handler for a
+// callback method. Note that none of the classes are pure; all reactions have a
+// default empty reaction so that the user class only needs to override those
+// classes that it cares about.
 
 /// \a ServerBidiReactor is the interface for a bidirectional streaming RPC.
 template <class Request, class Response>
 class ServerBidiReactor : public internal::ServerReactor {
  public:
+  // NOTE: Initializing stream_ as a constructor initializer rather than a
+  //       default initializer because gcc-4.x requires a copy constructor for
+  //       default initializing a templated member, which isn't ok for atomic.
+  // TODO(vjpai): Switch to default constructor and default initializer when
+  //              gcc-4.x is no longer supported
+  ServerBidiReactor() : stream_(nullptr) {}
   ~ServerBidiReactor() = default;
 
-  /// Do NOT call any operation initiation method (names that start with Start)
-  /// until after the library has called OnStarted on this object.
-
   /// Send any initial metadata stored in the RPC context. If not invoked,
   /// any initial metadata will be passed along with the first Write or the
   /// Finish (if there are no writes).
-  void StartSendInitialMetadata() { stream_->SendInitialMetadata(); }
+  void StartSendInitialMetadata() {
+    auto* stream = stream_.load(std::memory_order_acquire);
+    if (stream == nullptr) {
+      grpc::internal::MutexLock l(&stream_mu_);
+      stream = stream_.load(std::memory_order_acquire);
```

Comment:
The second load for double-checked locking should be `relaxed` because it's synchronized with mutex.  The first loads/stores should be acquire/release pairs.